File: backend/app/data_sources/fred.py
from app.config import get_settings

INDICATORS = {
    "DFF": "Federal Funds Rate",
    "CPIAUCSL": "Consumer Price Index",
    "UNRATE": "Unemployment Rate",
    "GDP": "Gross Domestic Product",
    "T10Y2Y": "10Y-2Y Treasury Spread",
    "VIXCLS": "VIX Volatility Index",
    "DGS10": "10-Year Treasury Rate",
}


def fetch_indicators() -> dict:
    """Fetch key economic indicators from FRED."""
    settings = get_settings()
    if not settings.FRED_API_KEY:
        return {}

    try:
        from fredapi import Fred

        fred = Fred(api_key=settings.FRED_API_KEY)
        results = {}
        for series_id, name in INDICATORS.items():
            try:
                series = fred.get_series(series_id)
                latest = series.dropna().iloc[-1]
                results[series_id] = {
                    "name": name,
                    "value": round(float(latest), 4),
                    "date": str(series.dropna().index[-1].date()),
                }
            except Exception:
                continue
        return results
    except Exception:
        return {}


def fetch_risk_free_rate() -> float | None:
    """Fetch the latest 10-Year Treasury rate from FRED (series DGS10).

    Returns rate as a decimal (e.g., 0.042 for 4.2%), or None if unavailable.
    """
    settings = get_settings()
    if not settings.FRED_API_KEY:
        return None
    try:
        from fredapi import Fred

        fred = Fred(api_key=settings.FRED_API_KEY)
        series = fred.get_series("DGS10")
        latest = series.dropna().iloc[-1]
        return round(float(latest) / 100, 4)
    except Exception:
        return None

File: laravel-backend/app/DataSources/Fred.php
<?php

namespace App\DataSources;

use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\Log;

class Fred
{
    private const INDICATORS = [
        'DFF' => 'Federal Funds Rate',
        'CPIAUCSL' => 'Consumer Price Index',
        'UNRATE' => 'Unemployment Rate',
        'GDP' => 'Gross Domestic Product',
        'T10Y2Y' => '10Y-2Y Treasury Spread',
        'VIXCLS' => 'VIX Volatility Index',
        'DGS10' => '10-Year Treasury Rate',
    ];

    private const BASE_URL = 'https://api.stlouisfed.org/fred';

    public function fetchIndicators(): array
    {
        $apiKey = config('portfolio.fred_api_key');
        if (empty($apiKey)) {
            return [];
        }

        $results = [];
        foreach (self::INDICATORS as $seriesId => $name) {
            try {
                $response = Http::timeout(10)->get(self::BASE_URL . '/series/observations', [
                    'series_id' => $seriesId,
                    'api_key' => $apiKey,
                    'file_type' => 'json',
                    'sort_order' => 'desc',
                    'limit' => 5,
                ]);

                if ($response->successful()) {
                    $observations = $response->json('observations', []);
                    foreach ($observations as $obs) {
                        if ($obs['value'] !== '.' && is_numeric($obs['value'])) {
                            $results[$seriesId] = [
                                'name' => $name,
                                'value' => round((float) $obs['value'], 4),
                                'date' => $obs['date'],
                            ];
                            break;
                        }
                    }
                }
            } catch (\Throwable $e) {
                continue;
            }
        }

        return $results;
    }

    public function fetchRiskFreeRate(): ?float
    {
        $apiKey = config('portfolio.fred_api_key');
        if (empty($apiKey)) {
            return null;
        }

        try {
            $response = Http::timeout(10)->get(self::BASE_URL . '/series/observations', [
                'series_id' => 'DGS10',
                'api_key' => $apiKey,
                'file_type' => 'json',
                'sort_order' => 'desc',
                'limit' => 5,
            ]);

            if ($response->successful()) {
                $observations = $response->json('observations', []);
                foreach ($observations as $obs) {
                    if ($obs['value'] !== '.' && is_numeric($obs['value'])) {
                        return round((float) $obs['value'] / 100, 4);
                    }
                }
            }
        } catch (\Throwable $e) {
            // Silently fail
        }

        return null;
    }
}
